### PR TITLE
feat: initial restructure of training script

### PR DESCRIPTION
**Pull request by othertea**
# initial restructure of training script

## Overview

The purpose of this script is to do a fairly minimal restructuring of our training script. My main goal is to make it easier for multiple contributors to contribute to different aspects of model training. Most contributors should no longer have to modify `protein_lm/modeling/scripts/train.py`, which should reduce the occurrence of merge conflicts. For example, someone working on the training dataset should ideally be working on different files than someone who is adding a positional encoding scheme. (I didn't quite achieve this goal as they would both have to modify the config yaml.)  My secondary objective for this restructuring is to make it easier to follow the code, which also facilitates contributions.

The main function of train.py has been refactored by simply calling functions that get the relative components, e.g., `get_dataset(...), get_model(...)` instead of constructing them itself. The arguments to these functions ultimately come from a single yaml file that gets parsed into [pydantic](https://docs.pydantic.dev/latest/) BaseModels, providing a central place to specify all such arguments. Such an example yaml can be found in `protein_lm/configs/train/toy_localcsv.yaml`.

## Discussion

I don't have strong preferences on where these new files should go. Currently, partly because I did not want to move any other files, they are in `protein_lm/modeling/getters/`, but perhaps a better place is one level higher. For instance, we probably want some of these getters for evaluation. I'm very open to suggestions here -- as stated above, the main purpose of this particular PR is to make it easer for more people to contribute. We can also move these files in a later PR, which should be easier than dealing with a lot of merge conflicts.

Similarly, I am open to specifying these configs in other ways, instead of the current system with yamls and pydantic models. This current system could be improved within these frameworks but we could also consider others. For instance, we could use [hydra](https://hydra.cc) to both make it easier to override config options via the command line as well as to combinatorially combine components.

And of course, I'm open to suggestions for any other restructuring that should be a part of this PR.
 
## Yet to do

What this PR doesn't do (and doesn't plan to do):
- Major changes to the logic of the training code. I mostly kept it as a refactor/restructure.
- Command line overrides for configs. Currently, the way of trying different parameters is to create a new yaml file with different training options. 
- Refactor of the model definition. I didn't have time to get to this, but hopefully even just the restructure in this PR is helpful.
- I have verified that the train script runs without error and appears to train, but I have not yet tested it with a reasonably sized dataset. I.e., there could still be bugs in the training logic. I'm leaving this to a follow-up PR, as the scope of that task is larger than just a restructure.

### CODE REVIEW

**Comment by jamaliki** [comment]
I'm a big fan of YAML config files, it makes everything reproducible and easy to track. Thank you so much for cleaning this up, @othertea

File: protein_lm/modeling/getters/training_args.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+
+def get_training_args(config_dict: Dict) -> TrainingArguments:
+    config = TrainingArgsConfig(**config_dict)
+
+    if not os.path.isdir(config.output_dir):
+        print(f"creating checkpoint directory at {config.output_dir}")
+        os.makedirs(config.output_dir)
+
+    training_args = TrainingArguments(
```
Comment (jamaliki) [inline protein_lm/modeling/getters/training_args.py:41]: Small nit, why can't this also be like the dataset arguments, etc? And also why the separation between TrainingArgsConfig and TrainingArguments (from Huggingface)? Seems like the API is consistent

File: protein_lm/modeling/getters/training_args.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        output_dir=config.output_dir,
+        max_steps=config.max_steps,
+        num_train_epochs=config.num_train_epochs,
+        per_device_train_batch_size=config.per_device_train_batch_size,
+        learning_rate=config.learning_rate,
+        weight_decay=config.learning_rate,
```
Comment (jamaliki) [inline protein_lm/modeling/getters/training_args.py:47]: also FYI you probably meant `weight_decay=config.weight_decay`

**Review by jamaliki — CHANGES_REQUESTED** [review]
Small nits

File: protein_lm/modeling/getters/training_args.py (hunk 1 above)
Comment (othertea) [reply protein_lm/modeling/getters/training_args.py:41]: I think you're raising some good points, though I'm not 100% sure I know what you're suggesting. Mainly it wasn't a very thoroughly-planned decision. 

* That there are two hard-coded arguments is just an oversight that I'll fix.
* At some point while writing this code, I wanted the validation that existed in the previous code for checking whether some params are nonnegative to remain in this version, since the purpose was a refactor. But upon reconsideration, I don't think this is important.
* I actually thought it would be more consistent to have a pydantic model (`TrainingArgsConfig`) rather than to just directly use huggingface's `TrainingArguments`, which appears to be just a dataclass. For instance, if we later decide to make a `TrainingConfig` pydantic model whose fields are all of the pydantic models I've introduced in this PR. But this is also not that important and could be changed later if it comes to that.

So in summary, if what you're suggesting is that I just return `TrainingArguments(**config_dict)` (and the accompanying changes to code) and get rid of `TrainingArgsConfig` entirely, I'm happy to make that change. Let me know if you meant something else!

File: protein_lm/modeling/getters/training_args.py (hunk 2 above)
Comment (othertea) [reply protein_lm/modeling/getters/training_args.py]: Thanks for catching that! It's fixed now.

File: protein_lm/modeling/getters/training_args.py (hunk 1 above)
Comment (jamaliki) [reply protein_lm/modeling/getters/training_args.py:41]: You also make good points. Having looked at the HF code's internals in the past few months, I would not be surprised if we would have to diverge from their conventions. It would also make it simper later on if we decide to use the flash attention models, for example.

I'm convinced now, I think the pydantic model is good and that we should go head with it.

Thanks again for all your work!

**Review by jamaliki — APPROVED** [review]
LGTM!

File: protein_lm/modeling/getters/training_args.py (hunk 1 above)
Comment (othertea) [reply protein_lm/modeling/getters/training_args.py]: Sounds good to me! I really don't have a strong preference either way, but of course happy to keep it as is.

**Review by NZ99 — APPROVED** [review]
looks great to me too!

**Review by jeffreyruffolo — APPROVED** [review]
Looks great to me!
